### PR TITLE
feats: support flexible KL methods, logits returns, and custom completion masks

### PR DESCRIPTION
**Pull request by Yuxuan-Zhang-Dexter**
This PR introduces enhancements to `tunix/rl/common.py`

### CODE REVIEW

**Comment by Yuxuan-Zhang-Dexter** [comment]
@googlebot rescan

> Thanks for your pull request! It looks like this may be your first contribution to a Google open source project. Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
> View this [failed invocation](https://github.com/google/tunix/pull/265/checks?check_run_id=49818125633) of the CLA check for more information.
> 
> For the most up to date status, view the checks section at the bottom of the pull request.

**Comment by yuxuandexter** [comment]
@googlebot rescan

File: tunix/rl/common.py [hunk 1]
```diff
@@ -100,23 +100,47 @@ class TrainExample:
 
 
 def compute_kl_divergence(
-    per_token_logps: jax.Array, ref_per_token_logps: jax.Array
+    per_token_logps: jax.Array, 
+    ref_per_token_logps: jax.Array, 
+    method: str | None = None
 ) -> jax.Array:
   """Compute per token KL divergence between trained and reference policy.
+  kl: Unbiased, high-variance estimator - simple forward KL: logp - ref_logp
```
Comment (abheesht17) [inline tunix/rl/common.py]: nit: A blank line after the first doc-string line

Comment (abheesht17) [reply tunix/rl/common.py]:
```
kl: Unbiased, high-variance estimator - simple forward KL: logp - ref_logp
```

to

```
Based of `method`, we compute one of three kinds of KL divergence:

`"kl"`: Unbiased, high-variance estimator. Simple forward KL: `logp - ref_logp`
```

Good to do this below as well.

File: tunix/rl/common.py [hunk 2]
```diff
@@ -100,23 +100,47 @@ class TrainExample:
...
+  mse_kl: Biased, low-variance estimator - squared log-difference: 0.5 * (logp - ref_logp)^2
+  low_var_kl: Unbiased, low-variance estimator - J. Schulman low-variance approx: (r - 1) - log r, where r = q/p = exp(ref_logp - logp)
+
+  references:
+  J. Schulman KL approx: http://joschu.net/blog/kl-approx.html
+  Verl code: https://github.com/volcengine/verl/blob/f9035b70166b102de5e84c1819e2172b64a0ae30/verl/trainer/ppo/core_algos.py#L1323-L1358
```
Comment (abheesht17) [inline tunix/rl/common.py]: Let's pin the code reference to a release tag.

File: tunix/rl/common.py [hunk 3]
```diff
@@ -100,23 +100,47 @@ class TrainExample:
...
 
   Args:
     per_token_logps: Per token log probabilities from the trained policy.
     ref_per_token_logps: Per token log probabilities from the reference policy.
+    method: KL penalty method. If None, defaults to "low_var_kl"..
```
Comment (abheesht17) [inline tunix/rl/common.py]: Remove extra period at the end

File: tunix/rl/common.py [hunk 4]
```diff
@@ -100,23 +100,47 @@ class TrainExample:
...
+
+  if method == "mse_kl":
+    # Biased but low variance.
+    return 0.5 * jnp.square(delta)
+
+  if method == "low_var_kl":
+    # Default to low-variance, unbiased estimator.
+    # (r - 1) - log(r) = exp(ref_logp - logp) - (ref_logp - logp) - 1
+    return jnp.exp(ref_per_token_logps - per_token_logps) \
+           - (ref_per_token_logps - per_token_logps) - 1
```
Comment (abheesht17) [inline tunix/rl/common.py]: Should we reuse delta here?

File: tunix/rl/common.py [hunk 5]
```diff
@@ -100,23 +100,47 @@ class TrainExample:
 
 
 def compute_kl_divergence(
-    per_token_logps: jax.Array, ref_per_token_logps: jax.Array
+    per_token_logps: jax.Array, 
+    ref_per_token_logps: jax.Array, 
+    method: str | None = None
```
Comment (abheesht17) [inline tunix/rl/common.py]: Remove `None`, change to

```
method: str = "low_var_kl"
```

File: tunix/rl/common.py [hunk 6]
```diff
@@ -100,23 +100,47 @@ class TrainExample:
...
   Returns:
     KL divergence.
   """
-  per_token_kl = (
-      jnp.exp(ref_per_token_logps - per_token_logps)
-      - (ref_per_token_logps - per_token_logps)
-      - 1
-  )
-  return per_token_kl
+  delta = per_token_logps - ref_per_token_logps
```
Comment (abheesht17) [inline tunix/rl/common.py]: Let's add a check up here...

```
if method not in (...):
  raise ValueError("`method` must be one of ... Received: ...")
```

File: tunix/rl/common.py [hunk 7]
```diff
@@ -100,23 +100,47 @@ class TrainExample:
...
+
+  if method == "kl":
+    # Unbiased but high variance.
```
Comment (abheesht17) [inline tunix/rl/common.py]: Don't need these comments, we already have this up in the doc-string

File: tunix/rl/common.py [hunk 8]
```diff
@@ -135,6 +159,7 @@ def selective_log_softmax(logits: jax.Array, input_ids: jax.Array) -> jax.Array:
 
 
 # TODO(tsbao): remove this once old callsite is cleaned up.
+
```
Comment (abheesht17) [inline tunix/rl/common.py]: Revert this change?

File: tunix/rl/common.py [hunk 9]
```diff
@@ -151,6 +176,30 @@ def get_per_token_logps(
   input_tokens = input_tokens[:, -logits_to_keep:]
   return selective_log_softmax(logits, input_tokens)
 
+# Need raw logits for entropy regularization
+@nnx.jit(static_argnums=(4,))
+def get_per_token_logps_and_logits(
```
Comment (abheesht17) [inline tunix/rl/common.py]: Can we modify the existing get_per_token_logps_and_logits function, with an additional `return_logits`? We return a tuple of `(per_token_logps, logits_slice)` if True, otherwise, just `per_token_logps`. What do you think?

**Review by abheesht17 — CHANGES_REQUESTED** [review]
Thanks for the PR! Reviewed some parts of it

File: tunix/rl/common.py (hunk 4 above)
Comment (yuxuandexter) [reply tunix/rl/common.py]: For a clarify, I decoupled the delta by directly computing `per_token_logps - ref_per_token_logps`. We can reuse delta in low_var_kl but '-delta' may look weird.

File: tunix/rl/common.py (hunk 9 above)
Comment (yuxuandexter) [reply tunix/rl/common.py]: I think that you mean the existing `get_per_token_logps` function, right? Adding a flag `return_logits" will be a cleaner alternative. We can do so.

**Comment by abheesht17** [comment]
Could you also run this: https://github.com/google/tunix/blob/main/code_style.sh?

**Comment by yuxuandexter** [comment]
> Could you also run this: https://github.com/google/tunix/blob/main/code_style.sh?

Nice tool! I have checked and reformated `tunix/rl/common.py`, `tunix/rl/inference/inference_worker.py`, `tunix/rl/rl_cluster.py`, `tunix/rl/rollout/vanilla_rollout.py`. Formatted with Pyink using 2-space indentation and an 80-character line length.

File: tunix/rl/common.py [hunk 10]
```diff
@@ -100,23 +100,42 @@ class TrainExample:
...
+    per_token_logps: jax.Array,
+    ref_per_token_logps: jax.Array,
+    method: str = "low_var_kl",
 ) -> jax.Array:
   """Compute per token KL divergence between trained and reference policy.
 
+  Based of `method`, we compute one of three kinds of KL divergence:
+  "kl": Unbiased, high-variance estimator. Simple Forward KL: `logp - ref_logp`
+  "mse_kl": Biased, low-variance estimator. Squared log-difference: `0.5 * (logp - ref_logp)^2`
+  "low_var_kl": Unbiased, low-variance estimator. J. Schulman low-variance approx: `(r - 1) - log r`, where `r = q/p = exp(ref_logp - logp)`
```
Comment (abheesht17) [inline tunix/rl/common.py]: Did the linter not flag this line? This looks longer than 80 characters.

File: tunix/rl/common.py [hunk 11]
```diff
@@ -100,23 +100,42 @@ class TrainExample:
...
 
 def compute_kl_divergence(
-    per_token_logps: jax.Array, ref_per_token_logps: jax.Array
+    per_token_logps: jax.Array,
+    ref_per_token_logps: jax.Array,
+    method: str = "low_var_kl",
 ) -> jax.Array:
   """Compute per token KL divergence between trained and reference policy.
 
+  Based of `method`, we compute one of three kinds of KL divergence:
```
Comment (abheesht17) [inline tunix/rl/common.py]: My bad, Based of  --> Based on

File: tunix/rl/common.py [hunk 12]
```diff
@@ -200,18 +231,52 @@ def compute_per_token_logps(
   return per_token_logps
 
 
-@nnx.jit(static_argnames=('pad_id', 'eos_id', 'stop_gradient'))
+@nnx.jit(static_argnames=("pad_id", "eos_id", "stop_gradient"))
+def compute_per_token_logps_and_logits(
```
Comment (abheesht17) [inline tunix/rl/common.py]: Can we fold this into `compute_per_token_logps` as well?

File: tunix/rl/common.py [hunk 13]
```diff
@@ -135,37 +154,48 @@ def selective_log_softmax(logits: jax.Array, input_ids: jax.Array) -> jax.Array:
...
   """Processes prompt and completion ids."""
 
   prompt_completion_ids = jnp.concat([prompt_tokens, completion_tokens], axis=1)
   prompt_mask = prompt_tokens != pad_id
-  completion_mask = make_completion_mask(completion_tokens, eos_tok=eos_id)
+  if completion_mask is None:
+    completion_mask = make_completion_mask(completion_tokens, eos_tok=eos_id)
+  else:
+    # Normalise dtype/values to int32 {0,1}
+    completion_mask = (completion_mask > 0).astype(jnp.int32)
```
Comment (abheesht17) [inline tunix/rl/common.py]: Just one quick question...

Do we need to do `(completion_mask > 0)`? Can't we use the passed `completion_mask` directly?

File: tunix/rl/common.py [hunk 14]
```diff
@@ -200,18 +231,52 @@ def compute_per_token_logps(
...
-@nnx.jit(static_argnames=('pad_id', 'eos_id', 'stop_gradient'))
+@nnx.jit(static_argnames=("pad_id", "eos_id", "stop_gradient"))
+def compute_per_token_logps_and_logits(
+    model: nnx.Module,
+    prompt_tokens: jax.Array,
+    completion_tokens: jax.Array,
+    pad_id: int,
+    eos_id: int,
+    stop_gradient: bool = True,
+    completion_mask: jax.Array | None = None,
```
Comment (abheesht17) [inline tunix/rl/common.py]: Let's move completion_mask to before `stop_gradient` in the arg order. Let's do this in other fns too

File: tunix/rl/common.py (hunk 10 above)
Comment (yuxuandexter) [reply tunix/rl/common.py]: No changes were made. I double-checked by running `pyink --line-length 80 --pyink-indentation=2 tunix/rl/common.py` directly, and also by deleting and re-pasting to run the `code_style.sh` script. 
Both approaches should have applied the same formatting, but the file remained unchanged.

File: tunix/rl/common.py (hunk 13 above)
Comment (yuxuandexter) [reply tunix/rl/common.py]: We can directly use completion mask. I will modify it.

**Review by abheesht17 — APPROVED** [review]
